warehouse/query-core/src/test/java/datawave/query/tld/TLDTest.java

```diff
@@ -0,0 +1,265 @@
+        ByteSequence grandchild = new ArrayByteSequence(grandchildId);
+        assertTrue(TLD.isRootPointer(root));
+        assertFalse(TLD.isRootPointer(child));
+        assertFalse(TLD.isRootPointer(grandchild));
+    }
+    
+    @Test
+    public void testFromString() {
+        ByteSequence expected = new ArrayByteSequence(rootId);
+        ByteSequence fromString = TLD.fromString(rootId);
+        assertEquals(expected, fromString);
+    }
+    
+    @Test
+    public void testParseParentPointerFromId() {
+        ByteSequence root = new ArrayByteSequence(rootId);
+        ByteSequence child = new ArrayByteSequence(childId);
+        ByteSequence grandchild = new ArrayByteSequence(grandchildId);
+        
+        assertEquals(root, TLD.parseParentPointerFromId(root));
+        assertEquals(root, TLD.parseParentPointerFromId(child));
+        assertEquals(root, TLD.parseParentPointerFromId(grandchild));
+    }
+    
+    @Test
+    public void testInstancesOf() {
+        ByteSequence bytes = new ArrayByteSequence("how.many.dots.do.we.have?");
+        ArrayList<Integer> expectedPositions = Lists.newArrayList(3, 8, 13, 16, 19);
+        ArrayList<Integer> instances = TLD.instancesOf('.', bytes);
+        assertEquals(expectedPositions, instances);
+    }
+    
+    @Test
+    public void testInstancesOfWithRepeat() {
+        ByteSequence bytes = new ArrayByteSequence("how.many.dots.do.we.have?");
+        ArrayList<Integer> expectedPositions = Lists.newArrayList(3, 8, 13);
+        ArrayList<Integer> instances = TLD.instancesOf('.', bytes, 3);
+        assertEquals(expectedPositions, instances);
+    }
+    
+    @Test
+    public void testLastInstancesOf() {
+        ByteSequence bytes = new ArrayByteSequence("how.many.dots.do.we.have?");
+        ArrayList<Integer> expectedPositions = Lists.newArrayList(19, 16, 13, 8, 3);
+        ArrayList<Integer> instances = TLD.lastInstancesOf('.', bytes, -1);
+        assertEquals(expectedPositions, instances);
+    }
+    
+    @Test
+    public void testLastInstancesOfWithRepeat() {
+        ByteSequence bytes = new ArrayByteSequence("how.many.dots.do.we.have?");
+        ArrayList<Integer> expectedPositions = Lists.newArrayList(19, 16, 13);
+        ArrayList<Integer> instances = TLD.lastInstancesOf('.', bytes, 3);
+        assertEquals(expectedPositions, instances);
+    }
+    
+    // Build parent keys from FI keys
+    @Test
+    public void testBuildParentKey() {
+        Key fiKey = buildFiKey(rootId);
+        Text shard = fiKey.getRow();
+        ByteSequence id = new ArrayByteSequence(rootId);
+        ByteSequence cq = fiKey.getColumnQualifierData();
+        Text cv = fiKey.getColumnVisibility();
+        long ts = fiKey.getTimestamp();
+        
+        Key parentKey = TLD.buildParentKey(shard, id, cq, cv, ts);
+        String cq2 = value + '\u0000' + datatype + '\u0000' + rootId;
+        Key expectedKey = new Key(row, rootId, cq2, fiKey.getTimestamp());
```

Comment:
You could use your `ts` variable for these `fiKey.getTimestamp()` calls.

Reply:
+1